go-controller/pkg/util/multi_network.go

```diff
@@ -60,20 +73,346 @@ type BasicNetInfo interface {
+	GetNetInfo() NetInfo
 }
 
-// NetInfo correlates which NADs refer to a network in addition to the basic
-// network information
-type NetInfo interface {
-	BasicNetInfo
-	GetNADs() []string
-	HasNAD(nadName string) bool
+// DefaultNetInfo is the default network information
+type DefaultNetInfo struct {
+	mutableNetInfo
+}
+
+// MutableNetInfo is a NetInfo where selected information can be changed.
+// Intended to be used by network managers that aggregate network information
+// from multiple sources that can change over time.
+type MutableNetInfo interface {
+	NetInfo
+
+	// NADs referencing a network
 	SetNADs(nadName ...string)
 	AddNADs(nadName ...string)
 	DeleteNADs(nadName ...string)
+
+	// VRFs a pod network is being advertised on, also per node
+	SetPodNetworkAdvertisedVRFs(podAdvertisements map[string][]string)
+
+	// Nodes advertising Egress IP
+	SetEgressIPAdvertisedVRFs(eipAdvertisements map[string][]string)
+}
+
+// NewMutableNetInfo builds a copy of netInfo as a MutableNetInfo
+func NewMutableNetInfo(netInfo NetInfo) MutableNetInfo {
+	if netInfo == nil {
+		return nil
+	}
+	return copyNetInfo(netInfo).(MutableNetInfo)
+}
+
+// ReconcilableNetInfo is a NetInfo that can be reconciled
+type ReconcilableNetInfo interface {
+	NetInfo
+
+	// canReconcile checks if both networks are compatible and thus can be
+	// reconciled. Networks are compatible if they are defined by the same
+	// static network configuration.
+	canReconcile(NetInfo) bool
+
+	// needsReconcile checks if both networks hold differences in their dynamic
+	// network configuration that could potentially be reconciled. Note this
+	// method does not check for compatibility.
+	needsReconcile(NetInfo) bool
+
+	// reconcile copies dynamic network configuration information from the
+	// provided network
+	reconcile(NetInfo)
+}
+
+// NewReconcilableNetInfo builds a copy of netInfo as a ReconcilableNetInfo
+func NewReconcilableNetInfo(netInfo NetInfo) ReconcilableNetInfo {
+	if netInfo == nil {
+		return nil
+	}
+	return copyNetInfo(netInfo).(ReconcilableNetInfo)
+}
+
+// AreNetworksCompatible checks if both networks are compatible and thus can be
+// reconciled. Networks are compatible if they are defined by the same
+// static network configuration.
+func AreNetworksCompatible(l, r NetInfo) bool {
+	if l == nil && r == nil {
+		return true
+	}
+	if l == nil || r == nil {
+		return false
+	}
+	return reconcilable(l).canReconcile(r)
+}
+
+// DoesNetworkNeedReconciliation checks if both networks hold differences in their dynamic
+// network configuration that could potentially be reconciled. Note this
+// method does not check for compatibility.
+func DoesNetworkNeedReconciliation(l, r NetInfo) bool {
+	if l == nil && r == nil {
+		return false
+	}
+	if l == nil || r == nil {
+		return true
+	}
+	return reconcilable(l).needsReconcile(r)
+}
+
+// ReconcileNetwork reconciles the dynamic network configuration
+func ReconcileNetwork(to ReconcilableNetInfo, from NetInfo) error {
```

Comment:
can we make this more specific, like ReconcileNetInfo ? ReconcileNetwork in other controllers looks like its actually configuring something in the system and its really not. Or maybe change to "UpdateNetworkState" or something.